### PR TITLE
DHCP unit test

### PR DESCRIPTION
**Pull request by AniruddhaKanhere**
Description
-----------
This PR adds unit-tests for the FreeRTOS_DHCP file.
With this PR the line, function, and branch coverage is at 100%.

Test Steps
-----------

Related Issue
-----------

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: FreeRTOS_ARP.c [hunk 1]
```diff
@@ -725,17 +725,16 @@ void FreeRTOS_OutputARPRequest( uint32_t ulIPAddress )
 
         #if defined( ipconfigETHERNET_MINIMUM_PACKET_BYTES )
             {
-                if( pxNetworkBuffer->xDataLength < ( size_t ) ipconfigETHERNET_MINIMUM_PACKET_BYTES )
```
Comment (htibosch) [inline FreeRTOS_ARP.c:728]: We talked about this if() clause: it must stay there. You have corrected that in the PR for FreeRTOS_ARP.c, and the change shouldn't be here.

**Review by htibosch — APPROVED** [review]
Except for the changes in FreeRTOS_ARP.c ( please revert ), there are no essential changes to the FreeRTOS+TCP library sources. So I can approve that part.

File: test/unit-test/FreeRTOS_DHCP/FreeRTOS_DHCP_utest.c [hunk 2]
```diff
@@ -0,0 +1,3657 @@
...
+    {
+        /* Modify the global state. */
+        xDHCPData.eDHCPState = i;
+        eReturn = eGetDHCPState();
+        TEST_ASSERT_EQUAL( i, eReturn );
+    }
+}
+
+NetworkBufferDescriptor_t * pxGlobalNetworkBuffer[ 10 ];
+uint8_t GlobalBufferCounter = 0;
```
Comment (alfred2g) [inline test/unit-test/FreeRTOS_DHCP/FreeRTOS_DHCP_utest.c:64]: statics... a few variables down the line as well

File: test/unit-test/FreeRTOS_DHCP/FreeRTOS_DHCP_utest.c [hunk 3]
```diff
@@ -0,0 +1,3657 @@
...
+        /* And this too. */
+        xDHCPData.ulDHCPServerAddress = 0xABABABAB;
+
+
+        /* Make random number generation pass. */
+        xApplicationGetRandomNumber_ExpectAndReturn( &( xDHCPData.ulTransactionId ), pdTRUE );
+        /* Return a valid socket. */
+        FreeRTOS_socket_ExpectAndReturn( FREERTOS_AF_INET, FREERTOS_SOCK_DGRAM, FREERTOS_IPPROTO_UDP, &xTestSocket );
+        /* Ignore the inputs to setting the socket options. */
+        FreeRTOS_setsockopt_IgnoreAndReturn( pdPASS );
```
Comment (alfred2g) [inline test/unit-test/FreeRTOS_DHCP/FreeRTOS_DHCP_utest.c:255]: better have them as ExpectAnyArgsAndReturn / ExpectAndReturn if they take no arguments

File: test/unit-test/FreeRTOS_DHCP/FreeRTOS_DHCP_utest.c [hunk 4]
```diff
@@ -0,0 +1,3657 @@
...
```
Comment (alfred2g) [inline test/unit-test/FreeRTOS_DHCP/FreeRTOS_DHCP_utest.c:301]: same here and a few down the line

File: test/unit-test/FreeRTOS_DHCP/FreeRTOS_DHCP_utest.c [hunk 5]
```diff
@@ -0,0 +1,3657 @@
...
+        vDHCPProcess( pdTRUE, i );
+
+        /* Expected state is incorrect, but we are trying to reset
+         * the DHCP the state machine. */
+        TEST_ASSERT_EQUAL( eWaitingSendFirstDiscover, xDHCPData.eDHCPState );
+        TEST_ASSERT_EQUAL( xDHCPSocket, &xTestSocket );
+        TEST_ASSERT_EQUAL( xDHCPData.xUseBroadcast, 0 );
+        /* This should be reset as well */
+        TEST_ASSERT_EQUAL( xDHCPData.ulOfferedIPAddress, 0 );
+        /* And this too. */
```
Comment (alfred2g) [inline test/unit-test/FreeRTOS_DHCP/FreeRTOS_DHCP_utest.c:316]: should the the other way arround

TEST_ASSERT_XXX ( expected_value, variable_to_test ); as doing it the other way around makes the message output by unity on failure look weird 

there are a few other occurrences of this in this file.

File: test/unit-test/FreeRTOS_DHCP/FreeRTOS_DHCP_utest.c [hunk 6]
```diff
@@ -0,0 +1,3657 @@
...
+    TEST_ASSERT_EQUAL( eInitialWait, xDHCPData.eDHCPState );
+    /* make sure that the period is increased by a factor of two. */
+    TEST_ASSERT_EQUAL( ( ( ipconfigMAXIMUM_DISCOVER_TX_PERIOD >> 1 ) - 1 ) << 1, xDHCPData.xDHCPTxPeriod );
+}
+
+static int32_t FreeRTOS_recvfrom_eWaitingOfferRecvfromSucceedsFalseCookieNoTimeout( Socket_t xSocket,
+                                                                                    void * pvBuffer,
+                                                                                    size_t uxBufferLength,
+                                                                                    BaseType_t xFlags,
+                                                                                    struct freertos_sockaddr * pxSourceAddress,
```
Comment (alfred2g) [inline test/unit-test/FreeRTOS_DHCP/FreeRTOS_DHCP_utest.c:891]: cleaner if these are at the top... if the file is too large, maybe better to split it

File: include/FreeRTOSIPConfigDefaults.h [hunk 7]
```diff
@@ -396,6 +396,7 @@
 #endif
 
 #if ( ipconfigDHCP_FALL_BACK_AUTO_IP != 0 )
+    #undef ipconfigARP_USE_CLASH_DETECTION
```
Comment (htibosch) [inline include/FreeRTOSIPConfigDefaults.h:399]: MISRA checkers do not like the use of `#undef`. Is there another solution?

**Review by htibosch — DISMISSED** [review]
I approve all changes that were made to the core sources ( `FreeRTOS_DHCP.{c,h}` ).
Thank you!

File: test/unit-test/FreeRTOS_DHCP/FreeRTOS_DHCP_utest.c (hunk 2 above)
Comment (AniruddhaKanhere) [reply test/unit-test/FreeRTOS_DHCP/FreeRTOS_DHCP_utest.c]: Done.

File: test/unit-test/FreeRTOS_DHCP/FreeRTOS_DHCP_utest.c (hunk 3 above)
Comment (AniruddhaKanhere) [reply test/unit-test/FreeRTOS_DHCP/FreeRTOS_DHCP_utest.c]: Agreed, done.

File: test/unit-test/FreeRTOS_DHCP/FreeRTOS_DHCP_utest.c (hunk 4 above)
Comment (AniruddhaKanhere) [reply test/unit-test/FreeRTOS_DHCP/FreeRTOS_DHCP_utest.c]: Done

File: test/unit-test/FreeRTOS_DHCP/FreeRTOS_DHCP_utest.c (hunk 6 above)
Comment (AniruddhaKanhere) [reply test/unit-test/FreeRTOS_DHCP/FreeRTOS_DHCP_utest.c]: I concur. Done.

File: include/FreeRTOS_DHCP.h [hunk 8]
```diff
@@ -103,6 +225,8 @@
 /* Internal call: returns true if socket is the current DHCP socket */
     BaseType_t xIsDHCPSocket( Socket_t xSocket );
 
+    struct xDHCPMessage_IPv4;
+
```
[the remark this answers is not in the document]
Comment (AniruddhaKanhere) [reply include/FreeRTOS_DHCP.h]: Yes, that was a mistake. Corrected.

File: include/FreeRTOSIPConfigDefaults.h (hunk 7 above)
Comment (AniruddhaKanhere) [reply include/FreeRTOSIPConfigDefaults.h:399]: Without `#undef`, we'll have a warning saying that this macro was re-defined. Any suggestions?

**Review by alfred2g — DISMISSED** [review]

File: include/FreeRTOSIPConfigDefaults.h (hunk 7 above)
Comment (htibosch) [reply include/FreeRTOSIPConfigDefaults.h:399]: Here are the default definition:
~~~c
#ifndef ipconfigDHCP_FALL_BACK_AUTO_IP
    #define ipconfigDHCP_FALL_BACK_AUTO_IP     0
#endif

#if ( ipconfigDHCP_FALL_BACK_AUTO_IP != 0 )
    #define ipconfigARP_USE_CLASH_DETECTION    1
#endif

#ifndef ipconfigARP_USE_CLASH_DETECTION
    #define ipconfigARP_USE_CLASH_DETECTION    0
#endif
~~~
which means that all you have to do is define `ipconfigDHCP_FALL_BACK_AUTO_IP` in your FreeRTOSIPConfig.h

Comment (htibosch) [reply include/FreeRTOSIPConfigDefaults.h:399]: I pressed the green button too quickly.

Have you defined `ipconfigARP_USE_CLASH_DETECTION` in your FreeRTOSIPConfig.h file?

What about testing as follows:
~~~c
#if ( ipconfigDHCP_FALL_BACK_AUTO_IP != 0 )
    #ifndef ipconfigARP_USE_CLASH_DETECTION
        #define ipconfigARP_USE_CLASH_DETECTION    1
    #else
        #if( ipconfigARP_USE_CLASH_DETECTION != 1 )
            #error ipconfigARP_USE_CLASH_DETECTION should be defined as 1 when AUTO_IP is used.
        #endif
    #endif
#endif
~~~
Would that work for you project?

File: include/FreeRTOSIPConfigDefaults.h [hunk 9]
```diff
@@ -396,8 +396,13 @@
...
-    #undef ipconfigARP_USE_CLASH_DETECTION
-    #define ipconfigARP_USE_CLASH_DETECTION    1
+    #ifndef ipconfigARP_USE_CLASH_DETECTION
+        #define ipconfigARP_USE_CLASH_DETECTION    1
+    #else
+        #if ( ipconfigARP_USE_CLASH_DETECTION != 1 )
+            #error ipconfigARP_USE_CLASH_DETECTION should be defined as 1 when AUTO_IP is used.
+        #endif
+    #endif
 #endif
```
Comment (htibosch) [inline include/FreeRTOSIPConfigDefaults.h:406]: All right, I am also happy with that, thank you.

**Review by htibosch — COMMENTED** [review]
Thank you for these last changes.
Hein

**Review by alfred2g — APPROVED** [review]

**Review by n9wxu — APPROVED** [review]

**Pull request merged**